sql: make JSON.size() take string header and slice header into account

heyihong opened the pull request:
@justinj. For calculating json array size, we should take the size of slice header into account. For calculating json object size, we should not count the size of string header twice. Some reference: https://blog.golang.org/go-slices-usage-and-internals

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21614)
<!-- Reviewable:end -->

knz commented:
Hello @heyihong ! Thank you again for your contribution.

Here I notice that the change pertains to a bug fix -- the behavior was previously incorrect, and you are changing this. Therefore, the `release note` text should mention the bug fix.

justinj commented:
Same here - will merge after addressing knz's remark!

heyihong committed:
sql: make JSON.size() take string header and slice header into account

Release note (bug fix): fixed inaccurate JSON.size() calculation for
jsonArray and jsonObject

heyihong commented:
Done. Please take another look.

justinj commented:
Looks great, thanks!

Amruta-Ranade commented:
As discussed with Ben, excluding this update from the release notes since this is an internal change.
